scanpy/neighbors/__init__.py

```diff
@@ -768,66 +507,80 @@ def compute_neighbors(
         if n_neighbors > self._adata.shape[0]:  # very small datasets
             n_neighbors = 1 + int(0.5 * self._adata.shape[0])
             logg.warning(f'n_obs too small: adjusting to `n_neighbors = {n_neighbors}`')
-        if method == 'umap' and not knn:
-            raise ValueError('`method = \'umap\' only with `knn = True`.')
-        if method not in {'umap', 'gauss', 'rapids'}:
-            raise ValueError("`method` needs to be 'umap', 'gauss', or 'rapids'.")
+        method, transformer_cls, shortcut = self._handle_transform_args(
+            method, transformer_cls, knn=knn
+        )
         if self._adata.shape[0] >= 10000 and not knn:
             logg.warning('Using high n_obs without `knn=True` takes a lot of memory...')
         # do not use the cached rp_forest
         self._rp_forest = None
         self.n_neighbors = n_neighbors
         self.knn = knn
         X = _choose_representation(self._adata, use_rep=use_rep, n_pcs=n_pcs)
+
         # neighbor search
         use_dense_distances = (metric == 'euclidean' and X.shape[0] < 8192) or not knn
-        if use_dense_distances:
+        if shortcut and (use_dense_distances or X.shape[0] < 4096):
```

Comment:
Could we use the `sklearn.KNeighborsTransformer(algorithm="brute")` for this? At least when `knn=False`?

Reply:
trying